design/Pipelines.md

```diff
@@ -0,0 +1,277 @@
+# Pipeline objects
+
+This is a document with pseudo-code for the parts of the API related to pipeline objects.
+Naming and the UX of the API are just an example and could change, the important parts are what’s available and what goes where.
+For example this document uses C structures when a C++ API might want to use builder objects, and a Javascript API could use dictionaries.
+
+## Creating pipeline objects
+
+For type safety, compute and graphics pipeline are separate types.
+To create a pipeline, a structure containing all the relevant information is passed to ```DeviceCreate<TYPE>Pipeline```.
+
+## Compute pipeline creation
+
+To create a compute pipeline the only things needed are some shader code present in a ```ShaderModule``` object, and a ```PipelineLayout``` object describing how the pipeline interacts with the binding model.
+
+```cpp
+struct ComputePipelineDescriptor {
+    ShaderModule module;
+    const char* entryPoint;
+    PipelineLayout layout;
+};
+
+ComputePipeline CreateComputePipeline(Device device, const ComputePipelineDescriptor* descriptor);
+```
+
+Translation to the backing APIs would be the following:
+ - **D3D12**: Translates to ```ID3D12::CreateComputePipelineState```, a ```D3D12_SHADER_BYTECODE``` is created from the ```(module, entryPoint)``` pair, and the ```ID3D12RootSignature``` is equivalent to the ```PipelineLayout```.
+ - **Metal**: Translates to ```MTLDevice::makeComputePipelineState```, the ```MTLFunction``` is created from the ```(module, entryPoint, layout)``` tuple by adapting the generated MSL to the resource slot allocation done in ```layout```.
+ - **Vulkan**: Translates to ```vkCreateComputePipelines``` with one pipeline. The ```vkShaderStageInfo``` corresponds to ```(module, entryPoint)``` and the ```vkPipelineLayout``` corresponds to ```layout```.
+
+Open questions:
+How do we take advantage of the pipeline caching present in D3D12 and Vulkan?
+Do we expose it to the application or is it done magically in the WebGPU implementation?
+
+## Render pipeline creation
+
+Render pipelines need ```ShaderModule``` and a ```PipelineLayout``` like compute pipelines and in addition require information about:
+ - Layout for vertex inputs
+ - Layout for fragment outputs
+ - All the fixed-function state
+
+For simplicity we assume most fixed-function state is created in separate object.
+For example a ```DepthStencilState``` object would be allocated and a pointer to it would be stored in the ```RenderPipelineDescriptor```.
+
+Mismatch:
+ - Metal doesn’t have primitive restart.
```

Comment:
What Metal actually doesn't have is a way to *disable* the primitive restart functionality.